xtask/src/swift.rs

```diff
@@ -85,7 +100,92 @@ fn build_library() -> Result<()> {
+        .join(target)
+        .join(rel_type_dir)
+        .join("libmatrix_sdk_ffi.a"))
+}
+
+fn build_xcframework(release_mode: bool, rsync_path: Option<PathBuf>) -> Result<()> {
+    println!("-- Building for iOS [1/5]");
+    let ios_path = build_for_target("aarch64-apple-ios", release_mode)?;
+
+    println!("-- Building for macOS (Apple Silicon) [2/5]");
+    let darwin_arm_path = build_for_target("aarch64-apple-darwin", release_mode)?;
+    println!("-- Building for macOS (Intel) [3/5]");
+    let darwin_x86_path = build_for_target("x86_64-apple-darwin", release_mode)?;
+
+    println!("-- Building for iOS Simulator (Apple Silicon) [4/5]");
+    let ios_sim_arm_path = build_for_target("aarch64-apple-ios-sim", release_mode)?;
+    println!("-- Building for iOS Simulator (Intel) [5/5]");
+    let ios_sim_x86_path = build_for_target("x86_64-apple-ios", release_mode)?;
+
+    println!("-- Creating XCFramework");
+
+    let root_dir = workspace::root_path()?;
+    let generated_dir = root_dir.join("generated");
+    let headers_dir = generated_dir.join("headers");
+    create_dir_all(headers_dir.clone())?;
+
+    // # MacOS
+    let lipo_target_macos = generated_dir.join("libmatrix_sdk_ffi_macos.a");
+    println!("-- Running Lipo for macOS [1/4]");
+    cmd!(
+        "lipo -create {darwin_x86_path} {darwin_arm_path}
+        -output {lipo_target_macos}"
+    )
+    .run()?;
+
+    // # iOS Simulator
+    let lipo_target_sim = generated_dir.join("libmatrix_sdk_ffi_iossimulator.a");
+    println!("-- Running Lipo for iOS [2/4]");
+    cmd!(
+        "lipo -create {ios_sim_arm_path} {ios_sim_x86_path}
+        -output {lipo_target_sim}"
+    )
+    .run()?;
+
+    println!("-- Generate uniffi files [3/4]");
+    generate_uniffi(&darwin_x86_path, &generated_dir)?;
+
+    fs::rename(generated_dir.join("matrix_sdk_ffiFFI.h"), headers_dir.join("matrix_sdk_ffiFFI.h"))?;
+
+    println!("-- Generate MatrixSDKFFI.xcframework framework [4/4]");
+    let xcframework_path = generated_dir.join("MatrixSDKFFI.xcframework");
+    if xcframework_path.exists() {
+        fs::remove_dir_all(xcframework_path.as_path())?;
+    }
+
+    cmd!(
+        "xcodebuild -create-xcframework
+        -library {lipo_target_macos}
+        -headers {headers_dir}
+        -library {lipo_target_sim}
+        -headers {headers_dir}
+        -library {ios_path}
+        -headers {headers_dir}
+        -output {generated_dir}/MatrixSDKFFI.xcframework
+    "
+    )
+    .run()?;
+
+    // cleaning up the intermediate data
+    fs::remove_file(lipo_target_macos.as_path())?;
+    fs::remove_file(lipo_target_sim.as_path())?;
+    fs::remove_dir_all(headers_dir.as_path())?;
+
+    if let Some(path) = rsync_path {
+        println!("-- Rsync MatrixSDKFFI.xcframework to {path:?}");
+        cmd!("rsync -a --delete {generated_dir}/MatrixSDKFFI.xcframework {path}").run()?;
+        cmd!("rsync -a --delete {generated_dir}/swift/ {path}/Sources/MatrixRustSDK").run()?;
+    }
```

Comment:
What's the idea behind this, is it for copying that to an iOS device via SSH or similar?

Reply:
no. the process currently is to copy the results into the `../matrix-rust-sdk-components-swift/` folder and run `codegen` in the app that links that components-folder. It is a bit crude, but this step being done directly removes an extra step (that is currently encoded into the debug-version of the script).

Reply:
If it's just for local moving, why require `rsync` to be installed? Could just use the Rust fs or potentially `xshell` helpers.

Reply:
this is more or less a copy from the previous bash script. I don't know if there are specific reasons to use rsync rather than a usual move, though the original scripts did use `mv` elsewhere and I thought it might be important that it's rsync for some reason.


@stefanceriu can you enlighten us?

Reply:
I honestly don't remember the reason, something about things not getting copied across properly. This was in the begining so the problem might've well been somewhere else.

I tested it again just now with a plain `cp -Rf` and it works just fine. All we need it to do is recursivelly copy and override everything.